libraries/AP_RangeFinder/AP_RangeFinder_HC_SR04.cpp

```diff
@@ -0,0 +1,177 @@
+        // no change
+        return;
+    }
+
+    // detach last one
+    if (echo_pin) {
+        if (!hal.gpio->detach_interrupt(echo_pin)) {
+            GCS_SEND_TEXT(MAV_SEVERITY_WARNING,
+                          "HC_SR04: Failed to detach from pin %u",
+                          echo_pin);
+            // ignore this failure or the user may be stuck
+        }
+    }
+
+    echo_pin = params.pin;
+
+    if (!params.pin) {
+        // don't need to install handler
+        return;
+    }
+
+    // install interrupt handler on rising and falling edge
+    hal.gpio->pinMode(params.pin, HAL_GPIO_INPUT);
+    if (!hal.gpio->attach_interrupt(
+            params.pin,
+            FUNCTOR_BIND_MEMBER(&AP_RangeFinder_HC_SR04::irq_handler,
+                                void,
+                                uint8_t,
+                                bool,
+                                uint32_t),
+            AP_HAL::GPIO::INTERRUPT_BOTH)) {
+        // failed to attach interrupt
+        GCS_SEND_TEXT(MAV_SEVERITY_WARNING,
+                      "HC_SR04: Failed to attach to pin %u",
+                      (unsigned int)params.pin);
+        return;
+    }
+}
+
+
+
+/*
+   detect if an HC_SR04 rangefinder is connected. The only thing we
+   can do is check if the pin number is valid. If it is, then assume
+   that the device is connected
+*/
+bool AP_RangeFinder_HC_SR04::detect(AP_RangeFinder_Params &_params)
+{
+    if (_params.pin == -1) {
+        return false;
+    }
+    if (_params.stop_pin == -1) {
+        return false;
+    }
+    return true;
+}
+
+
+// interrupt handler for reading distance-proportional time interval
+void AP_RangeFinder_HC_SR04::irq_handler(uint8_t pin, bool pin_high, uint32_t timestamp_us)
+{
+    if (pin_high) {
+        pulse_start_us = timestamp_us;
+    } else {
+        irq_value_us = timestamp_us - pulse_start_us;
+    }
+}
+
+/*
+  update distance_cm
+ */
+void AP_RangeFinder_HC_SR04::update(void)
+{
+    // check if pin has changed and configure interrupt handlers if required:
+    check_pins();
+
+    if (!echo_pin || ! trigger_pin || echo_pin == -1 || trigger_pin == -1) {
+        // disabled (either by configuration or failure to attach interrupt)
+        state.distance_cm = 0.0f;
+        return;
+    }
+
+    // disable interrupts and grab state
+    void *irqstate = hal.scheduler->disable_interrupts_save();
+    const uint32_t value_us = irq_value_us;
+    irq_value_us = 0;
+    hal.scheduler->restore_interrupts(irqstate);
+
+    const uint32_t now = AP_HAL::millis();
+    if (value_us == 0) {
+        // no reading; check for timeout:
+        if (now - last_reading_ms > 1000) {
+            // no reading for a second - something is broken
+            state.distance_cm = 0.0f;
+        }
+    } else {
+        // gcs().send_text(MAV_SEVERITY_WARNING, "Pong!");
+        // a new reading - convert time to distance
+        state.distance_cm = value_us / 58.0f;  // 58 is from datasheet
```

Comment:
would a mult be better here in terms of cpu time?

Reply:
Yes!